javalib/src/main/scala/java/util/Base64.scala

```diff
@@ -0,0 +1,390 @@
+package java.util
+
+import java.io.{IOException, InputStream, OutputStream}
+import java.nio.charset.StandardCharsets
+import java.nio.{ByteBuffer, IntBuffer}
+
+// scalastyle:off return
+object Base64 {
+
+  private type ByteTable = Array[Byte]
+  private type IntTable = Array[Int]
+
+  private val chars = ('A' to 'Z') ++ ('a' to 'z') ++ ('0' to '9')
+  private val basicEncodeTable: ByteTable = (chars ++ Seq('+', '/')).map(_.toByte).toArray
+  private val urlSafeEncodeTable: ByteTable = (chars ++ Seq('-', '_')).map(_.toByte).toArray
+
+  private def decodeTable(encode: ByteTable): IntTable = {
+    val decode: IntTable = Array.fill(256)(-1)
+    encode.zipWithIndex.foreach{ case (b,i) => decode(b) = i}
+    decode('=') = -2
+    decode
+  }
+
+  private val basicDecodeTable: IntTable = decodeTable(basicEncodeTable)
+  private val urlSafeDecodeTable: IntTable = decodeTable(urlSafeEncodeTable)
+
+  private val mimeLineSeparators = Array[Byte]('\r','\n')
+  private val mimeLineLength = 76
+
+  private val basicEncoder = new Encoder(basicEncodeTable)
+  private val basicDecoder = new Decoder(basicDecodeTable, false)
+
+  private val mimeEncoder = new Encoder(basicEncodeTable, mimeLineLength, mimeLineSeparators)
+  private val mimeDecoder = new Decoder(basicDecodeTable, true)
+
+  private val urlSafeEncoder = new Encoder(urlSafeEncodeTable)
+  private val urlSafeDecoder = new Decoder(urlSafeDecodeTable, false)
+
+  // --------------------------------------------------------------------------
+
+  def getDecoder: Decoder = basicDecoder
+
+  def getEncoder: Encoder = basicEncoder
+
+  def getMimeDecoder: Decoder = mimeDecoder
+
+  def getMimeEncoder: Encoder = mimeEncoder
+
+  def getMimeEncoder(lineLength: Int, lineSeparator: Array[Byte]): Encoder =
+    new Encoder(basicEncodeTable, lineLength / 4 * 4, lineSeparator)
+
+  def getUrlDecoder: Decoder = urlSafeDecoder
+
+  def getUrlEncoder: Encoder = urlSafeEncoder
+
+  // --------------------------------------------------------------------------
+
+  class Decoder private[Base64](table: IntTable, ignoreInvalid: Boolean) {
+
+    def decode(src: Array[Byte]): Array[Byte] = {
+      val dst = ByteBuffer.allocate(dstMaxLength(src.length))
```

Comment:
Why does manipulating a byte buffer have significant costs, isn't it just a wrapper around an array with a few ints to keep track of position, remaining etc..?

I agree that for the method that takes both the src and the dst array as arguments the allocation is unnecessary, but then we'll have to do tricks to find out if the dst array is large enough to accomodate for everything, we'll have to either traverse the src twice, once for the check (as the src can contain all sorts of garbage that is legal in the case of mime), another time for the decoding. Or we can just start decoding and throw if we run out of space in the dst. Both are inefficient in a way.